File: catkin_ws/src/students/scripts/practice03.cpp
#include <opencv2/core.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/opencv.hpp>
#include <iostream>
#include <stdio.h>
#include <math.h>
using namespace cv;
using namespace std;

int k_size = 5; //Variable kernel size for the gaussian filter
int t1=5; // lower threshold
int t2=10; // upper threshold
double pi = 3.14159265358979323846;
double delta=0;
int ddepth=-1;
int hough_threshold = 100;

std::vector<cv::Vec2f> hough_lines(cv::Mat img_bin, float d_min, float d_max, int d_res, float theta_min,
                                   float theta_max, float theta_res, int threshold)
{
    int dist_n  = (int)ceil((d_max - d_min)/d_res);
    int theta_n = (int)ceil((theta_max - theta_min)/theta_res);
    int votes[dist_n][theta_n]={};
    float theta_values[theta_n];
    std::cout << "Bins for d=" << dist_n << "  bins for theta=" << theta_n<< std::endl;
    for(size_t i=0; i<theta_n; i++)
        theta_values[i] = theta_min + theta_res*i;
    //for(size_t i=0; i<dist_n; i++)
      //  for(size_t j=0; j<theta_n; j++)
        //    votes[i][j] = 0;
    
    for(size_t i=0; i < img_bin.rows; i++)
        for(size_t j=0; j < img_bin.cols; j++)
            if(img_bin.at<unsigned char>(i,j) == 255)
                for(size_t k=0; k<theta_n; k++)
                {
                    int d = (int)((j*cos(theta_values[k]) + i*sin(theta_values[k]) - d_min)/d_res);
                    if(d >= 0 && d < dist_n)  votes[d][k]++;
                }

    
    std::vector<cv::Vec2f> lines;
    for(size_t i=0; i<dist_n; i++)
        for(size_t j=0; j<theta_n; j++)
            if(votes[i][j] > threshold)
            {
                float d = i*d_res + d_min;
                float a = j*theta_res + theta_min;
                lines.push_back(cv::Vec2f(d,a));
            }
    return lines;
}

void draw_lines(cv::Mat& img, std::vector<cv::Vec2f>& lines)
{
    for(size_t i=0; i< lines.size(); i++)
    {
        float d = lines[i][0], theta = lines[i][1];
        cv::Point p1, p2;
        double a = cos(theta), b = sin(theta);
        double x0 = a*d, y0 = b*d;
        p1.x = round(x0 + 1000*(-b));
        p1.y = round(y0 + 1000*(a));
        p2.x = round(x0 - 1000*(-b));
        p2.y = round(y0 - 1000*(a));
        cv::line(img, p1, p2, cv::Scalar(0,255,0), 2, cv::LINE_AA);
    }
}

cv:: Mat final_supress(cv::Mat& T){
    cv::Mat F=T.clone();
    for(size_t i=1; i < T.rows-1; i++)
        for(size_t j=1; j < T.cols-1; j++){
            if(T.at<float>(i,j)==60&&(T.at<float>(i+1,j)==255||T.at<float>(i+1,j+1)==255||T.at<float>(i+1,j-1)==255||T.at<float>(i-1,j)==255||T.at<float>(i-1,j+1)==255||T.at<float>(i-1,j-1)==255||T.at<float>(i,j+1)==255||T.at<float>(i,j-1)==255 ))
                F.at<float>(i,j)=255;
            else if(T.at<float>(i,j)==60)
                F.at<float>(i,j)=0;
        }
    F.convertTo(F,CV_8UC1);
    return F.clone();
}

cv::Mat filter_gaussian(int k,float sigma){
    int siz=(int)k/2;
    float suma=0;
    float H[2*siz+1][2*siz+1]={};
    float arg=0;
    for(int i=0; i<= 2*siz; i++){
        for(int j=0; j<= 2*siz; j++){
            arg=(pow(i-siz,2)+pow(j-siz,2))/(2*pow(sigma,2));
            H[i][j]=exp(-arg)/(2*pi*pow(sigma,2));
            suma=suma+H[i][j];
        }
    }
    for(int i=0; i<= 2*siz; i++){
        for(int j=0; j<= 2*siz; j++){
            H[i][j]=H[i][j]/suma;
        }
    }
    //CV_32F  CV_8UC1
    return cv::Mat(2*siz+1, 2*siz+1, CV_32F, &H).clone();
}

cv::Mat threshold(cv::Mat& G){
    cv::Mat T=G.clone();
    for(size_t i=1; i < G.rows-1; i++)
        for(size_t j=1; j < G.cols-1; j++){
            if(G.at<float>(i,j) > (float)t1 && G.at<float>(i,j) <= (float)t2)
                T.at<float>(i,j)=60;
            else if(G.at<float>(i,j) > (float)t2)
                T.at<float>(i,j)=255;
        }
    return T.clone();
}

cv::Mat get_sobel_x_gradient(cv::Mat& img){
    float Gx[3][3]={-1/8.0,0,1/8.0,-2/8.0,0,2/8.0,-1/8.0,0,1/8.0};
    cv::Mat SobelX=Mat(3, 3, CV_32F, &Gx);
    cv::Mat img_filter;
    //cout<<SobelX<<endl;
    filter2D(img, img_filter, ddepth , SobelX, cv::Point(-1, -1), delta, BORDER_DEFAULT );
    //CV_32F  CV_8UC1
    return img_filter.clone();
}

cv::Mat get_sobel_y_gradient(cv::Mat& img){
    float Gy[3][3]={-1/8.0,-2/8.0,-1/8.0,0,0,0,1/8.0,2/8.0,1/8.0};
    cv::Mat SobelY=Mat(3, 3, CV_32F, &Gy);
    cv::Mat img_filter;
    //cout<<SobelY<<endl;
    filter2D(img, img_filter, ddepth , SobelY, cv::Point(-1, -1), delta, BORDER_DEFAULT );
    //CV_32F  CV_8UC1
    return img_filter.clone();
}

void mag_angle(cv::Mat& A, cv::Mat& Gm, cv::Mat& Ga){
    cv::Mat SobelX=get_sobel_x_gradient(A);
    cv::Mat SobelY=get_sobel_y_gradient(A);
    for(size_t i=0; i < SobelX.rows; i++)
        for(size_t j=0; j < SobelX.cols; j++){
            Gm.at<float>(i,j)=sqrt(pow(SobelX.at<float>(i,j),2)+pow(SobelY.at<float>(i,j),2));
            Ga.at<float>(i,j)=atan2( SobelY.at<float>(i,j), SobelX.at<float>(i,j) );
            if(Ga.at<float>(i,j)<0)
                Ga.at<float>(i,j)+=pi;
            Ga.at<float>(i,j)=Ga.at<float>(i,j)*180/pi;
        }
    //Ga.convertTo(Ga,CV_8UC1);
}

cv::Mat supress_non_maximum(cv::Mat& Gm, cv::Mat& Ga){
    cv::Mat G=Gm.clone();
    size_t di=0;
    size_t dj=0;
    for(size_t i=1; i < G.rows-1; i++)
        for(size_t j=1; j < G.cols-1; j++){
            if(Ga.at<float>(i,j)<=22 || Ga.at<float>(i,j) > 157){
                di=0;
                dj=1;
            }
            else if(Ga.at<float>(i,j)>22 && Ga.at<float>(i,j) <= 67){
                di=1;
                dj=1;
            }
            else if(Ga.at<float>(i,j)>67 && Ga.at<float>(i,j) <= 112){
                di=1;
                dj=0;
            }
            else{
                di=1;
                dj=-1;
            }
            if(Gm.at<float>(i,j) >= Gm.at<float>(i+di,j+dj) && Gm.at<float>(i,j) > Gm.at<float>(i-di,j-dj) )
                G.at<float>(i,j)=Gm.at<float>(i,j);
            else
                G.at<float>(i,j)=0;
    }
    return G.clone();
}

void on_threshold_changed(int, void*){}

int main(int, char**)
{
    cv::Mat frame,Gm,G,Ga,gray,img_filter,T;
    cv::Mat SobelX,SobelY,kernel,F, Hough;
    //cout<<Arr<<endl;
    float sigma=1.1;
    //--- INITIALIZE VIDEOCAPTURE
    VideoCapture cap;
    // open the default camera using default API
    cap.open(0);
    cap.set(CAP_PROP_FRAME_WIDTH, 640);//Setting the width of the video
    cap.set(CAP_PROP_FRAME_HEIGHT, 480);//Setting the height of the video//
    cv::namedWindow("Original");
    cv::createTrackbar("Umbral inferior Canny:", "Original", &t1, 12, on_threshold_changed);
    cv::createTrackbar("Umbral superior Canny:", "Original", &t2, 22, on_threshold_changed);
    cv::createTrackbar("Tamaño filtro gaussiano:", "Original", &k_size, 9, on_threshold_changed);
    cv::createTrackbar("Umbral número de votos:", "Original", &hough_threshold, 255, on_threshold_changed);
    // check if we succeeded
    if (!cap.isOpened()) {
        cerr << "ERROR! Unable to open camera\n";
        return -1;
    }
    //--- GRAB AND WRITE LOOP
    cout << "Start grabbing" << endl
        << "Press ESC to terminate" << endl;
    while(cap.isOpened())
    {
        // wait for a new frame from camera and store it into 'frame'
        cap.read(frame);
        // check if we succeeded
        if (frame.empty()) {
            cerr << "ERROR! blank frame grabbed\n";
            break;
        }
        gray = frame.clone();
        //gray=gray/255.0;
        cv::cvtColor(gray, gray, cv::COLOR_BGR2GRAY);
        // Initialize arguments for the filter
        gray.convertTo(gray,CV_32FC1); //Change the class type in order to work with float numbers

        kernel=filter_gaussian(k_size, sigma);
        filter2D(gray, img_filter, ddepth , kernel, cv::Point(-1, -1), delta, BORDER_DEFAULT );

        Gm=img_filter.clone(); //Construct Gm
        Ga=img_filter.clone(); //Construct Ga

        mag_angle(img_filter, Gm, Ga);

        G=supress_non_maximum(Gm, Ga);

        T=threshold(G);
        SobelX=get_sobel_x_gradient(img_filter);
        SobelY=get_sobel_y_gradient(img_filter);

        F=final_supress(T);

        cv::Mat F_clone=F.clone();
        //F_clone.convertTo(F_clone,CV_8U);
        cv::cvtColor(F_clone, F_clone, cv::COLOR_GRAY2RGB);

        //hough_lines(cv::Mat img_bin, float d_min, float d_max, int d_res, float theta_min,
        //float theta_max, float theta_res, int threshold)
        std::vector<cv::Vec2f> lines = hough_lines(F, 25, 800, 1, -M_PI, M_PI, 0.05, hough_threshold);

        draw_lines(F_clone, lines);

        draw_lines(frame, lines);

        std::cout << "Number of lines: " << lines.size() << std::endl;

        //imshow("Hough transform", F_clone);
        //imshow("Hough", Hough);

        //Change the classs for visualizing Images
        Gm.convertTo(Gm,CV_8UC1);
        Ga.convertTo(Ga,CV_8UC1);
        T.convertTo(T,CV_8UC1);
        //show live and wait for a key with timeout long enough to show images
        //cout <<"Filer "<< G.type() << " == " << CV_32F << endl;
        imshow("Original", frame);
        //imshow("Grey",gray/255.0);
        //imshow("G",G);
        imshow("Canny",F_clone);
        //imshow("F",F);
        //cout<<kernel<<endl;
        if (waitKey(30) == 27)
            break;
    }
    // the camera will be deinitialized automatically in VideoCapture destructor
    return 0;
}

File: catkin_ws/src/students/scripts/practice03.py
import cv2
import numpy
import math


#-----------------------------------FILTRO GAUSSIANO-----------------------------------------------------
def convolve2d(A, kernel):
    return cv2.convertScaleAbs(cv2.filter2D(A, cv2.CV_16S, kernel))
#--------------------------------------------------------------------------------------------------------


    
#------------------------------------------GENERAR EL KERNEL (Se proponen 4)-----------------------------
def get_gaussian_kernel(k,sigma):
    k = k//2
    H = numpy.zeros((2*k+1, 2*k+1))
    r,c = H.shape
    for i in range(r):
        for j in range(c):
            H[i,j] = 1.0/(2*math.pi*sigma*sigma)*math.exp(-((i-k)**2 + (j-k)**2)/(2*sigma*sigma))
    H /= numpy.sum(H)
    return H         
#-------------------------------------------------------------------------------------------------------



#------------------------------------------Filtro Sobel x , y and (x,y) --------------------------------
def get_sobel_x_gradient(A):# gradiente en x
    Gx = numpy.asarray([[-1, 0, 1],[-2, 0, 2],[-1, 0, 1]])/8.0
    return convolve2d(A, Gx)

def get_sobel_y_gradient(A):# Gradiente en y
    Gy = numpy.asarray([[-1, -2, -1],[0, 0, 0],[1, 2, 1]])/8.0
    return convolve2d(A, Gy)

def get_sobel_x_and_y_gradient(A):# Gradiente en y
    Gx = numpy.asarray([[-1, 0, 1],[-2, 0, 2],[-1, 0, 1]])/8.0
    Gy = numpy.asarray([[-1, -2, -1],[0, 0, 0],[1, 2, 1]])/8.0
    return convolve2d(A, Gx), convolve2d(A, Gy)
#-------------------------------------------------------------------------------------------------------


#------------------------------------------ Magnitud y angulo del gradiente ----------------------------
def get_sobel_mag_angle(A):
    Gx = get_sobel_x_gradient(A)
    Gy = get_sobel_y_gradient(A)
    Gm = numpy.zeros(Gx.shape)
    Ga = numpy.zeros(Gx.shape)
    r,c = Gx.shape
    for i in range(r):
        for j in range(c):
            Gm[i,j] = math.sqrt(Gx[i,j]**2 + Gy[i,j]**2)
            Ga[i,j] = math.atan2(Gy[i,j],Gx[i,j])
            if Ga[i,j] < 0:
                Ga[i,j] += math.pi
            Ga[i,j] = int(Ga[i,j]/math.pi*180)
    return Gm.astype(numpy.uint8), Ga.astype(numpy.uint8)
#-------------------------------------------------------------------------------------------------------


#------------------------------- Supresion de no maximos and Histeresis -------------------------------
def supress_non_maximum(Gm, Ga):
    G = numpy.zeros(Gm.shape)
    r,c = Gm.shape
    for i in range(1,r-1):
        for j in range(1,c-1):
            if Ga[i,j] <= 22 or Ga[i,j] > 157:
                di, dj = 0, 1
            elif Ga[i,j] > 22 and Ga[i,j] <= 67:
                di, dj = 1, 1
            elif Ga[i,j] > 67 and Ga[i,j] <= 112:
                di, dj = 1, 0
            else:
                di, dj = 1, -1
            if Gm[i,j] >= Gm[i+di, j+dj] and Gm[i,j] > Gm[i-di, j-dj]:
                G[i,j] = Gm[i,j] 
            else:
                G[i,j] = 0
    return G.astype(numpy.uint8)
    
def doble_Umbral(Gm,Umin,Umax):
    G = numpy.zeros(Gm.shape)
    r,c = Gm.shape
    for i in range(1,r-1):
        for j in range(1,c-1):
            if Gm[i,j] > Umax:
                G[i,j] = Gm[i,j] 
            elif Gm[i,j] >= Umin and Gm[i,j] <= Umax:
                if Gm[i+1, j] > 0  or Gm[i-1, j] > 0 :
                    G[i,j] = Gm[i,j] 
                elif Gm[i, j+1] > 0  or Gm[i, j-1] > 0 :
                    G[i,j] = Gm[i,j] 
                elif Gm[i+1, j+1] > 0  or Gm[i-1, j-1] > 0 :
                    G[i,j] = Gm[i,j] 
                else:
                    G[i,j] = 0
            else:
                G[i,j] = 0
    return G.astype(numpy.uint8)

def doble_Umbral_Saturado(Gm,Umin,Umax):
    G = numpy.zeros(Gm.shape)
    r,c = Gm.shape
    for i in range(1,r-1):
        for j in range(1,c-1):
            if Gm[i,j] > Umax:
                G[i,j] = 255
            elif Gm[i,j] >= Umin and Gm[i,j] <= Umax:
                if Gm[i+1, j] > 0  or Gm[i-1, j] > 0 :
                    G[i,j] = 220
                elif Gm[i, j+1] > 0  or Gm[i, j-1] > 0 :
                    G[i,j] = 220
                elif Gm[i+1, j+1] > 0  or Gm[i-1, j-1] > 0 :
                    G[i,j] = 220
                else:
                    G[i,j] = 0
            else:
                G[i,j] = 0
    return G.astype(numpy.uint8)
#-------------------------------------------------------------------------------------------------------




#------------------------------------------------ Filtro CANNY -----------------------------------------
def canny_border_detector(filtered, Umin, Umax):
    Gm, Ga = get_sobel_mag_angle(filtered)
    Gsnm = supress_non_maximum(Gm, Ga)
    G = doble_Umbral(Gsnm,Umin,Umax)
    return G
def canny_border_detector_Saturado(filtered, Umin, Umax):
    Gm, Ga = get_sobel_mag_angle(filtered)
    Gsnm = supress_non_maximum(Gm, Ga)
    G = doble_Umbral_Saturado(Gsnm,Umin,Umax)
    return G
#-------------------------------------------------------------------------------------------------------

#-------------------------------------------------------------------------------------------------------
#-----------------------------Transfomada Hough-------------------------------------------
    
def transformada_hough(img, umbral, d_min, d_res, theta_min, theta_max, theta_res):
    
    alto, ancho = img.shape
    
    d_max = int(math.sqrt(alto**2 + ancho**2))
    d_n = numpy.arange(d_min, d_max, d_res)
    theta_n = numpy.arange(theta_min, theta_max, theta_res)
    accumulator = numpy.zeros((len(d_n), len(theta_n)))

    for i in range(alto):
        for j in range(ancho):
            if img[i,j] > 0:
                for k in range(len(theta_n)):
                    d = int(((j * math.cos(theta_n[k]) + i * math.sin(theta_n[k])) - d_min)//d_res)
                    if(d >= 0 and d < len(d_n)):
                        accumulator[d, k] += 1
    lineas = []
    for i in range(len(d_n)):
        for j in range(len(theta_n)):
            if accumulator[i,j] > umbral:
                d = d_n[i]
                theta = theta_n[j]
                lineas.append((d, theta))
    return lineas 
#-------------------------------------------------------------------------------------------------------


#-------------------------------------------------------------------------------------------------------
#-----------------------------Trazar Lineas-------------------------------------------
    
def trazar_lineas(G, lines):
    img = numpy.zeros(G.shape)
    for x in range(0,len(lines)-1):
        rho, theta = lines[x]
        a = numpy.cos(theta)
        b = numpy.sin(theta)
        x0 = a*rho
        y0 = b*rho
        x1 = int(x0 + 1000*(-b))
        y1 = int(y0 + 1000*(a))
        x2 = int(x0 - 1000*(-b))
        y2 = int(y0 - 1000*(a))
        cv2.line(img, (x1, y1), (x2, y2), (255, 255, 255), 2)
    return img
#-------------------------------------------------------------------------------------------------------


def Constante():
    return 0

#-------------------------------------------------------------------------------------------------------
#-------------------------------------------------------------------------------------------------------

def main():

    #----------------------------------------
    #--------- Video ------------------------
    cap  = cv2.VideoCapture(0)
    
    #------ Kernel y umbrales ---------------
    kernel = get_gaussian_kernel(5,1)
    kernel_1 = get_gaussian_kernel(7,1)
    kernel_2 = get_gaussian_kernel(3,1)
    
    pv = cv2.imread("Prueba.png");
    Pv= cv2.cvtColor(pv, cv2.COLOR_BGR2GRAY)
    PV = numpy.zeros(pv.shape)
    
    Umin = 5
    Umax = 255
    #------ Porcentaje de escala ------------
    scale_percent = 30
    
    #------ Variable auxiliar ---------------
    aux1 = 0
    Umbral = 50
    cv2.namedWindow('Transformada Hough')
    cv2.createTrackbar('Umbral','Transformada Hough',Umbral, 100, Constante)
   #-----------------------------------------
   
    while cap.isOpened():
    
    
    
        #--------- Video -----------------------------------------------------------------
        ret, frame = cap.read()
        if not ret:
            print("Can't receive frame (stream end?). Exiting ...")
            break
       #------------------------------------------------------------------------------------
       
       #--------- Escala la imagen --------------------------------------------------------
        if aux1 == 0:
            width = int(frame.shape[1] * scale_percent / 100)
            height = int(frame.shape[0] * scale_percent / 100)
            dsize = (width, height)
            aux1 = 1
        output = cv2.resize(frame, dsize)
        output = cv2.cvtColor(output, cv2.COLOR_BGR2GRAY)
        #------------------------------------------------------------------------------------
        
        
        #--------- Escala la imagen --------------------------------------------------------
        if aux1 == 1:
            filtered = convolve2d(output, kernel)
            Gx = get_sobel_x_gradient(filtered)
            Gy = get_sobel_y_gradient(filtered)
            cv2.imshow("Filtro Sobel en x Gx", Gx)
            cv2.imshow("Filtro Sobel en y Gy", Gy)
        elif aux1 == 2:
            filtered = convolve2d(output, kernel)
            Gm, Ga = get_sobel_mag_angle(filtered )
            cv2.imshow("Angulo del gradiente Ga", Ga)
            cv2.imshow("Magnitud del gradiente Gm", Gm)
        elif aux1 == 3:
            filtered = convolve2d(output, kernel)
            Gm, Ga = get_sobel_mag_angle(filtered )
            cv2.imshow("Angulo del gradiente Ga", Ga)
            cv2.imshow("Magnitud del gradiente Gm", Gm)
        elif aux1 == 4:
            filtered = convolve2d(output, kernel)
            Gm, Ga = get_sobel_mag_angle(filtered )
            G = supress_non_maximum(Gm, Ga)
            cv2.imshow("Supresion de no maximos", G)
        elif aux1 == 4:
            filtered = convolve2d(output, kernel)
            Gm, Ga = get_sobel_mag_angle(filtered )
            G = supress_non_maximum(Gm, Ga)
            Gf = doble_Umbral(G,Umin,Umax)
            cv2.imshow("Histeresis", Gf)
        elif aux1 == 5:
            filtered = convolve2d(output, kernel)
            G = canny_border_detector(filtered,Umin,Umax)
            cv2.imshow("Filtro Kanny", G)       
        elif aux1 == 6:
            filtered = convolve2d(output, kernel)
            G = canny_border_detector_Saturado(filtered,Umin,Umax)
            cv2.imshow("Filtro Kanny Saturado", G) 
        elif aux1 == 7:
            lines = transformada_hough(Pv, 200, 0,3, -numpy.pi/2, numpy.pi/2, numpy.pi/180)
            PV = trazar_lineas(PV,lines)
            cv2.imshow("PruebaHough", PV)
        elif aux1 >= 8:
            filtered = convolve2d(output, kernel)
            G = canny_border_detector_Saturado(filtered,10,255)
            GG = numpy.zeros(G.shape)
            lines = transformada_hough(G, 80, 0,3, -numpy.pi/2, numpy.pi/2, numpy.pi/180)
            GG = trazar_lineas(PV,lines)
            cv2.imshow("Transformada Hough", GG)
            
        else:
            cv2.imshow("Video", output)
        
        if aux1 <= 9:
           aux1 += 1
        #------------------------------------------------------------------------------------
        
        
        #--------- Fin  del programa------------------------------------------------------------ 
        if cv2.waitKey(10) & 0xFF == 27:
           break
    cap.release()
    cv2.destroyAllWindows()

if __name__ == '__main__':
    main()
#-------------------------------------------------------------------------------------------------------

